pkg/antctl/raw/multicluster/join.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicluster
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	multiclusterv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/pkg/antctl/raw/multicluster/common"
+)
+
+type clusterOptions struct {
+	Namespace string `yaml:"namespace,omitempty"`
+	ClusterID string `yaml:"clusterID,omitempty"`
+	Secret    string `yaml:"secret,omitempty"`
+	APIServer string `yaml:"apiServer,omitempty"`
+}
+
+type JoinOptions struct {
+	LeaderClusterOpt clusterOptions `yaml:"leaderCluster"`
+	MemberClusterOpt clusterOptions `yaml:"memberCluster"`
+	ClusterSetID     string         `yaml:"clusterSetID"`
+	ConfigFile       string
+	Secret           *v1.Secret
+}
+
+type meta struct {
+	Kind       string `yaml:"kind"`
+	APIVersion string `yaml:"apiVersion"`
+}
+
+var joinOpts *JoinOptions
+
+func (o *JoinOptions) validateAndComplete() error {
+	if o.ConfigFile != "" {
+		raw, err := ioutil.ReadFile(o.ConfigFile)
+		if err != nil {
+			return err
+		}
+		typeMeta := &meta{}
+		if err := yamlUnmarshall(raw, typeMeta); err != nil {
+			return err
+		}
+		if typeMeta.Kind != "ClusterSetConfig" || typeMeta.APIVersion != "multicluster.antrea.io/v1alpha1" {
+			return fmt.Errorf("unknown apiVersion or kind in config: %s", o.ConfigFile)
+		}
+		if err := yamlUnmarshall(raw, &joinOpts); err != nil {
+			return err
+		}
+		if o.LeaderClusterOpt.Secret == "" {
+			o.Secret, err = unmarshallSecret(raw)
+			if err != nil {
+				return fmt.Errorf("failed to unmarshall Secret from config file: %v", err)
+			}
+			if o.Secret.Namespace != o.MemberClusterOpt.Namespace {
```

Comment:
Could we just set the Namespace to o.MemberClusterOpt.Namespace?

When the Secret manifest is generated by init or "create membertoken", the Namespace should not be set, as the command is running on the leader and does not know the member Namespace.

Reply:
Agree with @jianjuns , I didn't see you set the Namespace in the Secret when your generate the file, this step seems unnecessary. Maybe just assign it to o.MemberClusterOpt.Namespace.

Reply:
Assigned it to o.MemberClusterOpt.Namespace now.